src/OVAL/probes/unix/process58.c

```diff
@@ -351,12 +352,91 @@ static int get_exec_shield_status(int pid) {
 	return ret;
 }
 
+/**
+ * Parse /proc/%d/cmdline file
+ * @param filepath Path to file ~ use preallocated buffer for the path
+ * @param buffer output buffer with non-zero size
+ * @return ps-like command info or NULL
+ */
+static inline bool get_process_cmdline(const char* filepath, struct oscap_buffer* const buffer){
+
+	int fd = open(filepath, O_RDONLY, 0);
+
+	if (fd < 0) {
+		return false;
+	}
+
+	oscap_buffer_clear(buffer);
+
+
+
+	for(;;) {
+		static const int chunk_size = 1024;
+		char chunk[chunk_size];
+		// Read data, store to buffer
+		ssize_t read_size = read(fd, chunk, chunk_size );
+		oscap_buffer_append_binary_data(buffer, chunk, read_size);
+
+		// If reach end of file, then end the loop
+		if (chunk_size != read_size) {
+			break;
+		}
+	}
+
+	close(fd);
+
+	int length = oscap_buffer_get_length(buffer);
+	char* buffer_mem = oscap_buffer_get_raw(buffer);
+
+	if ( length == 0 ) { // empty file
+		return false;
+	} else {
+		/**
+		 * Skip multiple trailing zeros
+		 */
+		int i = length - 1;
+		while ( (i > 0) && (buffer_mem[i] == '\0') ) {
+			--i;
+		}
+
+		/**
+		 * Program and args are separated by '\0' and '\n'
+		 * Replace them with spaces ' '
+		 */
+		while( i >= 0 ){
+			char chr = buffer_mem[i];
+			if ( ( chr == '\0') || ( chr == '\n' ) ) {
+				buffer_mem[i] = ' ';
+			} else if ( chr == '\033' ) {
+				buffer_mem[i] = '.';
+			}
+			--i;
+		}
+	}
+	return true;
+}
+
+/**
+ * Make "[%s] <defunct>" from cmd string - inplace
+ * @param cmd_buffer @see read_process() > cmd_buffer
+ * @return pointer to start of string
+ */
+static inline char *make_defunc_str(char* const cmd_buffer){
+	static const char DEFUNC_STR[] = "] <defunct>";
+
+	size_t len = strlen(cmd_buffer);
+	memcpy(cmd_buffer + len, DEFUNC_STR, sizeof(DEFUNC_STR));
+	return cmd_buffer;
+}
+
 static int read_process(SEXP_t *cmd_ent, SEXP_t *pid_ent, probe_ctx *ctx)
 {
 	int err = 1;
 	DIR *d;
 	struct dirent *ent;
 
+	struct oscap_buffer *cmdline_buffer = oscap_buffer_new();
+
 	d = opendir("/proc");
 	if (d == NULL)
```

Comment:
What about oscap_buffer_free() here? I suggest moving `struct oscap_buffer *cmdline_buffer = oscap_buffer_new();` after this.